codec/gcm.go

```diff
@@ -1,26 +1,58 @@
 package codec
 
 import (
+	"bytes"
 	"crypto/aes"
 	"crypto/cipher"
-	"encoding/binary"
 	"fmt"
 	"io"
-	"math/rand"
+	"math"
+
+	"github.com/renproject/id"
+	"github.com/renproject/surge"
 )
 
+type gcmNonce struct {
+	top       uint32
+	bottom    uint64
+	countDown bool
+}
+
+func (nonce *gcmNonce) change() {
+	if nonce.countDown {
+		nonce.pred()
+	} else {
+		nonce.succ()
+	}
+
+}
+
+func (nonce *gcmNonce) succ() {
+	nonce.bottom += 1
+	if nonce.bottom == 0 {
```

Comment:
Again, a small comment explaining that `bottom` is "overflowing" into `top` would be very helpful.